src/net_processing.cpp

```diff
@@ -663,6 +663,24 @@ bool ProcessMessage(CNode *pfrom, std::string strCommand, CDataStream &vRecv, in
         // This step done after final handshake
         CheckAndRequestExpeditedBlocks(pfrom);
     }
+    else if (strCommand == NetMsgType::XUPDATE)
+    {
+        CXVersionMessage xUpdate;
+        vRecv >> xUpdate;
+        // check for peer trying to change non-changeable key
+        for (auto entry : xUpdate.xmap)
+        {
+            auto iter = XVer::mapKeyType.find(entry.first);
+            if (iter == XVer::mapKeyType.end())
+            {
+                continue;
+            }
+            else if (iter->second == XVer::keyType::changeable)
+            {
+                pfrom->xVersion.xmap[entry.first] = xUpdate.xmap[entry.first];
```

Comment:
xVersion didn't need a lock when it was initialized and never changed, but XUPDATE means we need a lock.

I think that we should make a fine-grained lock just for xVersion to not interfere with other tasks that threads may be doing with this node.

Reply:
one possibility would be to encapsulate the lock in CXVersionMessage, and to offer set and get member functions rather than exposing xmap.